### PR TITLE
drivers: counter: Add SAM0 basic counter support

### PR DESCRIPTION
**Pull request by Sizurka**
This adds support for the basic timer counter (TC) found on SAM0 series parts.  This driver only supports running the counter in 32 bit wide mode.  Since this mode explicitly slaves the odd counters to the even ones, only instances of the even ones are defined.

Tested with `tests/drivers/counter/counter_basic_api` on SAMD21.

Compile tested on SAMR21 and SAMD20. ~~but requires #15104 (and a bit of rebasing if that's merged)~~

This will also require a trivial rebase if #14794 is merged (the parameter is already there, just currently set to a default).

@benpicco: I compile tested this against SAME54, but you'd need to generate a `tc_fixup_samd5x.h` in the spirit of the sercom one.  I just used the static definition:
```c
#define MCLK_TC0 (&MCLK->APBAMASK.reg)
#define MCLK_TC0_MASK (MCLK_APBAMASK_TC0)
````

### CODE REVIEW

File: drivers/counter/counter_sam0_tc32.c [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+}
+
+static int counter_sam0_tc32_set_alarm(struct device *dev, u8_t chan_id,
+		const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct counter_sam0_tc32_data *data = DEV_DATA(dev);
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
+	TcCount32 *tc = cfg->regs;
+
+	if (chan_id != 0) {
```
Comment (nordic-krch) [inline drivers/counter/counter_sam0_tc32.c]: this is already checked in `counter_set_channel_alarm`.

File: drivers/counter/counter_sam0_tc32.c [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+		tc->INTENSET.reg = TC_INTFLAG_MC1;
+	} else {
+		counter_sam0_tc32_relative_alarm(dev, alarm_cfg->ticks);
+	}
+
+	irq_unlock(key);
+
+	return 0;
+
+err_busy:
```
Comment (nordic-krch) [inline drivers/counter/counter_sam0_tc32.c]: this seems unnecessary as it's used only in one place.

File: drivers/counter/counter_sam0_tc32.c [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+	return -EBUSY;
+}
+
+static int counter_sam0_tc32_cancel_alarm(struct device *dev, u8_t chan_id)
+{
+	struct counter_sam0_tc32_data *data = DEV_DATA(dev);
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
+	TcCount32 *tc = cfg->regs;
+
+	if (chan_id != 0) {
```
Comment (nordic-krch) [inline drivers/counter/counter_sam0_tc32.c]: it's checked in `counter_cancel_channel_alarm()`

File: drivers/counter/counter_sam0_tc32.c [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+	 * Unsync read is safe here because we're not using
+	 * capture mode, so things are only set from the CPU
+	 * end.
+	 */
+	return tc->CC[0].reg;
+}
+
+static u32_t counter_sam0_tc32_get_max_relative_alarm(struct device *dev)
+{
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
```
Comment (nordic-krch) [inline drivers/counter/counter_sam0_tc32.c]: what about `return counter_sam0_tc32_get_top_value(dev) - 1;`

**Review by nordic-krch — CHANGES_REQUESTED** [review]

**Comment by benpicco** [comment]
Nice work!
On SAME54 the odd timer has to be enabled separately. Luckily both even and odd timers map to the same GCLK_ID and they can be assumed to belong to the same APB group.

So all I had to do was to change `.mclk_mask = MCLK_TC##n##_MASK` to `.mclk_pos = MCLK_TC##n##_Pos`  and add `*cfg->mclk |= 1 << (cfg->mclk_pos + 1)`

With this, `tests/drivers/counter/counter_basic_api` runs successful on SAME54.

File: drivers/counter/counter_sam0_tc32.c (hunk 4 above)
Comment (Sizurka) [reply drivers/counter/counter_sam0_tc32.c]: Updated with all the requested changes.

**Comment by Sizurka** [comment]
Note that the Kconfig error is due to token pasting (`CONFIG_COUNTER_SAM0_TC32_##n##_DIVISOR`) so it's spurious.  Should I open a PR on the CI tools to add this symbol, even fore this is merged?

**Comment by benpicco** [comment]
needs a rebase

**Review by benpicco — APPROVED** [review]
tested both on samr21-xpro and same54-xpro

**Comment by benpicco** [comment]
I think the prescaler selection should go into the device tree - but I suppose you will touch up on this anyway once #15585 is merged.

**Comment by Sizurka** [comment]
Yeah, I'd do a `clock-frequency` property in the device tree for that.  My original logic was that the frequency of the counter would be better as application controllable (vs defined by hardware), which makes it fit Kconfig better.  

However, once the SoC clock tree is not an immutable thing, the prescaler should be configured the same way: meaning device tree specification and calculation derived from the requested frequency.

**Review by nordic-krch — APPROVED** [review]
looks ok

**Pull request merged**